zio-http/src/main/scala/zio/http/Server.scala

```diff
@@ -1,300 +1,62 @@
-
-  /**
-   * Launches the app on the provided port.
-   */
-  def start[R](port: Int, http: HttpApp[R, Throwable]): ZIO[R, Throwable, Nothing] =
-    start(new InetSocketAddress(port), http)
-
-  def start[R](address: InetAddress, port: Int, http: HttpApp[R, Throwable]): ZIO[R, Throwable, Nothing] =
-    start(new InetSocketAddress(address, port), http)
-
-  def start[R](socketAddress: InetSocketAddress, http: HttpApp[R, Throwable]): ZIO[R, Throwable, Nothing] =
-    (Server(http).withBinding(socketAddress).make *> ZIO.never)
-      .provideSomeLayer[R](EventLoopGroup.auto(0) ++ ServerChannelFactory.auto ++ Scope.default)
-
-  def unsafePipeline(pipeline: ChannelPipeline => Unit): UServer = UnsafeChannelPipeline(pipeline)
-
-  def unsafeServerBootstrap(serverBootstrap: ServerBootstrap => Unit): UServer = UnsafeServerBootstrap(serverBootstrap)
-
-  /**
-   * Holds server start information.
-   */
-  final case class Start(port: Int = 0)
-
-  private[zio] final case class Config[-R, +E](
-    leakDetectionLevel: LeakDetectionLevel = LeakDetectionLevel.SIMPLE,
-    error: Option[Throwable => ZIO[R, Nothing, Unit]] = None,
-    sslOption: ServerSSLOptions = null,
-
-    // TODO: move app out of settings
-    app: HttpApp[R, E] = Http.empty,
-    address: InetSocketAddress = new InetSocketAddress(8080),
-    acceptContinue: Boolean = false,
-    keepAlive: Boolean = true,
-    consolidateFlush: Boolean = false,
-    flowControl: Boolean = true,
-    channelInitializer: ChannelPipeline => Unit = null,
-    requestDecompression: (Boolean, Boolean) = (false, false),
-    objectAggregator: Int = 1024 * 100,
-    serverBootstrapInitializer: ServerBootstrap => Unit = null,
-  ) {
-    def useAggregator: Boolean = objectAggregator >= 0
-  }
-
-  sealed trait LeakDetectionLevel {
-    self =>
-    def jResourceLeakDetectionLevel: ResourceLeakDetector.Level = self match {
-      case LeakDetectionLevel.DISABLED => ResourceLeakDetector.Level.DISABLED
-      case LeakDetectionLevel.SIMPLE   => ResourceLeakDetector.Level.SIMPLE
-      case LeakDetectionLevel.ADVANCED => ResourceLeakDetector.Level.ADVANCED
-      case LeakDetectionLevel.PARANOID => ResourceLeakDetector.Level.PARANOID
-    }
-  }
-
-  object LeakDetectionLevel {
-    case object DISABLED extends LeakDetectionLevel
-
-    case object SIMPLE extends LeakDetectionLevel
-
-    case object ADVANCED extends LeakDetectionLevel
-
-    case object PARANOID extends LeakDetectionLevel
+  val test = ServerConfig.testConfig >>> live
+
+  private final case class ServerLive(
+                                       appRef: java.util.concurrent.atomic.AtomicReference[HttpApp[Any, Throwable]],
+                                       bindPort: Int
+                                     ) extends Server {
+    override def serve[R](httpApp: HttpApp[R, Throwable]): URIO[R, Unit] =
+      ZIO.environment[R].map { env =>
+        val newApp = httpApp.provideEnvironment(env)
```

Comment:
You can add a fast-path optimization here by doing something like:

```suggestion
        val newApp = if (env == ZEnvironment.empty) httpApp.asInstanceOf[HttpApp[Any, Throwable]] else httpApp.provideEnvironment(env)
```

Reply:
Done